java-checks/src/main/java/org/sonar/java/checks/security/CryptographicKeySizeCheck.java

```diff
@@ -0,0 +1,151 @@
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks.security;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import org.sonar.check.Rule;
+import org.sonar.java.model.LiteralUtils;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.tree.AssignmentExpressionTree;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.LiteralTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.NewClassTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+
+@Rule(key = "S4426")
+public class CryptographicKeySizeCheck extends IssuableSubscriptionVisitor {
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Collections.singletonList(Tree.Kind.METHOD);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (!hasSemantic()) {
+      return;
+    }
+    MethodVisitor methodVisitor = new MethodVisitor();
+    tree.accept(methodVisitor);
+  }
+
+  private class MethodVisitor extends BaseTreeVisitor {
+
+    private String algorithm = null;
+    private Symbol geInstanceKeyGenSymbol = null;
+
+    @Override
+    public void visitNewClass(NewClassTree tree) {
+      // skip anonymous classes
+    }
+
+    @Override
+    public void visitMethodInvocation(MethodInvocationTree tree) {
+      if (isKeyGenGetInstanceOrInitMethod(tree)) {
+        String methodName = tree.symbol().name();
+        ExpressionTree arg = tree.arguments().get(0);
+        if ("getInstance".equals(methodName)) {
+          if (arg.is(Tree.Kind.STRING_LITERAL)) {
+            extractSymbolOfAssignedInstance(tree).ifPresent(keyGenSymbol -> {
+              algorithm = LiteralUtils.trimQuotes(((LiteralTree) arg).value());
+              geInstanceKeyGenSymbol = keyGenSymbol;
+            });
+          }
+        } else if (isInitMethod(methodName) && initializeIsCalledOnPreviouslyDefinedKeyGen(tree)) {
+          Integer minKeySize = findMinimumKeySizeFromAlgorithm(algorithm);
+          if (minKeySize != -1) {
```

Comment:
The approach requires that getInstance is called only once in a method. Otherwise, we get some false negatives: that can be fine but would deserve some test cases to highlight the limits of the approach.